### ISSUE TITLE
[FR] Have an option to make `--bool`'s operators operate on whole file instead of lines

### ISSUE THREAD

**Issue by NightMachinary** [Genivia/ugrep#113]
`foo AND bar` searches for both `foo` and `bar` on the same line. What if we want to find a file that contains both these patterns?

Perhaps we can just repurpose `--dotall` for this; `AND` is kind of reminiscent of `.*`, after all.

**Comment by NightMachinary** [comment]
I have been thinking about this, and the more general feature is to simply have a `--read0` option that treats blocks separated by a NUL character as a single block to search in, instead of just assuming the blocks are separated by newlines.

**Comment by genivia-inc** [comment]
Thanks for the feedback. Grep is inherently line-based. But it would be an interesting and powerful option to add, indeed. Because the logic would apply to the entire file, either two scans over the file will be needed or the entire file should be buffered, because it may take until the end of the file to match all required AND patterns. It might be doable for a major release update.

I suppose this is best done with new options that specify additional requirements for file selections, not line selections. I will investigate.

**Comment by tylerw** [comment]
For comparison, [git-grep(1)](https://git-scm.com/docs/git-grep) has this feature and spells it `--all-match`:

```
--all-match
When giving multiple pattern expressions combined with --or, this flag is specified to limit the match to files that have lines to match all of them.
```

...and the man page includes an example for clarity:
```
git grep --all-match -e NODE -e Unexpected
Looks for a line that has NODE or Unexpected in files that have lines that match both.
```

**Comment by genivia-inc** [comment]
A new feature to select files that match multiple patterns is in the planning phase.

The `--all-match` option is interesting, though it is not a intuitive choice of option naming IMO. Hence their explanation with an example to clarify? It is not clear if this means that files are selected only when `NODE` and `Unexpected` match a line that match *both* while showing matches for either? Or does this select files that match `NODE` and `Unexpected` when both are found *somewhere* in the file? Both scenarios make sense, depending on what the user wants and what results he/she expects without unexpected surprises...

**Comment by NightMachinary** [comment]
What was wrong with adding an option to change the default delimiter from newlines to an arbitrary character? Implementing file search on top of that is trivial (just cat the files and print the delimiter between them).

**Comment by genivia-inc** [comment]
>  (just cat the files and print the delimiter between them)

Nothing wrong with adding `--read0` *in principle*, but this can lead to memory issues when the file searched is a couple of GB in size to still fit in memory (a line must fit in memory to report a match on that line). So this is still in a planning phase to figure out how to handle GBs or consider an alternative approach. Basically, we don't want any arbitrary limits which we have right now (except that each `\n`-terminated line cannot be longer than 2GB or so in order to report it as a match, which makes sense, IMHO.) Technically it's not as simple as replacing `\n` with `\0` in the input, because we still need to report line numbers and column offsets on a line-by-line basis. To ensure performance, files to search should preferably *never* be completely stored in memory, unless they are small. Instead, an optimal search window is used to search the files for performance and to ensure that (huge) file sizes do not matter at all.

There is a more clever way to find files that match Boolean AND-patterns, although we still should expect to have to store an entire file in memory in the worst case (rather than efficiently streaming the file with a small window size to search). The worst case is when `A AND B` matches an `A` at the start of the file but there is no `B` in the file. In this case we have to store the entire file in memory but discard it. Note that the worst case is also almost always the most common case, because we do not expect many files to match. As a consequence, there will be a lot of pressure on memory usage when searching large files, whether or not they match. And it gets much worse when we have more terms to match, e.g. `A AND B AND C ... AND Z` will likely match one of the terms in a file at the start, but not necessarily all of the terms.

The reason I bring this up is that I doubt that other tools like git-grep consider these considerations and decisions important, therefore leaving it to the user to find out that the search does not work with large GB files or is very slow in general.

**Comment by tylerw** [comment]
> It is not clear if this means that files are selected only when NODE and Unexpected match a line that match both while
> showing matches for either? Or does this select files that match NODE and Unexpected when both are found somewhere
> in the file?

Yes, I agree it is ambiguous. From a quick test, it's the latter that is meant. And as you mentioned, there are significant performance implications that the user is probably unaware of (until they hit a case where it matters).

**Comment by genivia-inc** [comment]
Perhaps this is reasonably understandable:
~~~
    --overall
            Apply Boolean queries globally to files, the opposite of --local.
            A file matches if all Boolean query conditions can be satisfied.
            For example, --overall -e A --and -e B matches a file if some lines
            match A and some lines match B.  May also be formulated as
            --overall --bool 'A B'.  See also options --and, --andnot, --not,
            --bool and --local.
    --local
            Apply Boolean queries locally to lines, the opposite of --overall.
            This is the default Boolean query mode to match specific lines.
~~~

**Comment by genivia-inc** [comment]
Alternative:
~~~
    --files
            Apply Boolean queries to match files, the opposite of --lines.  A
            file matches if all Boolean query conditions are satisfied by
            matching lines anywhere in the file.  For example, --files -e A
            --and -e B matches a file if some lines match `A' and some lines
            match `B'.  This may also be formulated as --files --bool 'A B'.
            See also options --and, --andnot, --not, --bool and --lines.
    --lines
            Apply Boolean queries to match lines, the opposite of --files.
            This is the default Boolean query mode to match specific lines.
~~~
1. Boolean queries can be arbitrarily complex (not just `AND` of two or more patterns.)

2. Git grep `--all-match` is a confusing option, because `--all-match -e A --or -e B` matches lines with `A` or `B` as long as BOTH `A` AND `B` occur in the file somewhere. Using the `--or` here isn't helping to explain what this option does. Because `--all-match` uses `--or`, ugrep should not introduce an `--all-match` that uses `--and` to do the same. IMO `--and` makes logically sense to match both `A` and `B` in the file anywhere.

**Comment by NightMachinary** [comment]
@genivia-inc Thanks! I like `--files` and `--lines`; `--file-oriented`, `--line-oriented` are also nice. The other options are confusing IMO.

**Comment by genivia-inc** [comment]
Yeah, I agree `--files` and `--lines` are better.

**Comment by genivia-inc** [comment]
Finished implementing `--files` with `--bool` (also the `--and`, `--andnot` and `--not` operators).

The good news is that there are no limitations when using `--files` with `--bool`. Arbitrary complex queries can be formulated.

There is no real bad news, except perhaps that the `-v` (`--invert-match`) command-line option cannot be combined with `--files`, because inverting matches applies specifically to lines, i.e. exclude matching lines to report non-matching lines. Not only does `-v` not apply to files. when we want something like this applied to files then the equivalent would be to negate (invert) the logical condition specified with `--bool` to produce the list of non-matching files.

Even though the basics of the approach are straightforward as I've explained before, I ran into some minor implementation challenges when `--files` is combined with some of the output-related options `-l`, `-c`, `-o`, `-y`, `--format`. These issues were eventually resolved in a clear, concise and consistent way to implement the additional logic for `--files` to produce correct results for all possible combinations.

I still want to spend some more time running more tests to see if I can break anything.
